appinventor/components/src/com/google/appinventor/components/runtime/ChartView.java

```diff
@@ -0,0 +1,286 @@
+        // * Using FutureTasks, Hardware Acceleration off (to make invalidate non-async) and run on UI thread
+        //   and FutureTask.get() to wait for refreshing to finish
+        // * Using FutureTasks and cancelling current Refresh tasks, and starting another one (+ CountDownLatch
+        //   approach combined with FutureTask cancelling)
+        // * Waiting for Refresh Tasks to finish via posted FutureTasks on the UI Handler
+        // The chosen solution is to then have delays and refresh throttling.
+
+
+        // An AtomicReference is used to hold a single Refresh runnable, which is then
+        // executed in a UIHandler. Since the AtomicReference holds a single Runnable
+        // instance, it also acts as an accumulator in the case of too many refresh
+        // calls being invoked within a time frame of 100ms.
+//        refreshRunnable.set(getRefreshRunnable());
+
+        // Post a Refresh runnable on the UI Thread (via the UI Handler),
+        // since refreshing should only be invoked in the UI thread (due
+        // to accessing views). A delay of 100ms is used to throttle the
+        // refresh rate to prevent crashes.
+//        uiHandler.postDelayed(new Runnable() {
+//            @Override
+//            public void run() {
+//                // Get the runnable from the AtomicReference
+//                Runnable runnable = refreshRunnable.getAndSet(null);
+//
+//                // Runnable non-null; Execute it
+//                if (runnable != null) {
+//                    runnable.run();
+//                }
+//            }
+//        }, 100);
+
+        // Alternate solution (very similar results):
+        //        if (!delayQueue.isEmpty()) {
+        //            return;
+        //        }
+        //
+        //        final long startTime = System.currentTimeMillis() + 100;
+        //
+        //        Delayed delay = new Delayed() {
+        //            @Override
+        //            public long getDelay(@NonNull TimeUnit unit) {
+        //                return unit.convert(startTime - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+        //            }
+        //
+        //            @Override
+        //            public int compareTo(@NonNull Delayed o) {
+        //                return 0;
+        //            }
+        //        };
+        //
+        //        delayQueue.put(delay);
+        //
+        //        try {
+        //            delayQueue.take();
+        //        } catch (InterruptedException e) {
+        //            e.printStackTrace();
+        //        }
+        //
+        //        chart.getData().notifyDataChanged();
+        //        chart.notifyDataSetChanged();
+        //        chart.postInvalidate();
+    }
+
+    /**
+     * Updates the specified Chart Data Model and refreshes the
+     * Chart.
+     * @param model  Chart Data Model to update & refresh
+     */
+    public void Refresh(final ChartDataModel model) {
```

Comment:
@ewpatton The underlying issue here lies a bit in the way that the Charts operate in the library. I have realized at some point that the key point of the issue is that when the refresh call happens in the UI, and something also happens in the background (i.e. entries are being added), the refresh call will process through the added entries in the background, which ultimately results in a crash. So the idea behind the workaround I implemented relies on the principle of having an immutable list (well, in the sense that it's never modified through the script) which gets replaced on every refresh, to avoid these sorts of conflicts.
I tried catching exceptions with regards to this as well, but unfortunately there are some exceptions that are out of scope of the exception catcher.. (they seem to occur in a place that was not reachable with a try .. catch).

I removed the long comments; That was more of a detailed description for the reviewing process to see if we can do better here. :) Would you still think we should try the proposed solution in such a case, since I think for the resulting answer now it would not make much a difference (except maybe performance wise)